Use bookworm/debian 12, python 3.12, update python deps, fix buildx warnings

fredvd opened the pull request:
Fix buildx warning for ARG and ENV usage.
update relstorage, python-ldap, add sasl2 for ldap installation. 
Fix libtiff5 to 6
Update and pin pip and wheel
Remove importlibmetadata env var for older pip issue. 
Fix FROM/as casing warning from buildx.

fredvd committed:
Use bookworm-slim, python3.12
Fix buildx warning for ARG and ENV usage.
update relstorage, python-ldap, add sasl2 for ldap installation.
Fix libtiff5 to 6
Update and pin pip and wheel
Remove importlibmetadata env var for older pip issue.
Fix FROM/as casing warning from buildx.

fredvd commented:
@mauritsvanrees @davisagli  The images build for me now locally (arm64), without warnings.  But when I try to start plone-backend or plone-classicui, the image halts because python cannot import Zope2.  

Comparing against local builds for 6.0.14, I notice that plone-backend is missing the path to site-packages, which explains why Zope2 cannot be found. 

But also my plone-backend:6.1.0rc1  local image still has Python 3.11, not 3.12 :-O

6.0.14 plone-backend

```
root@e8eadcbb1a86:/app# python3
Python 3.11.11 (main, Dec  4 2024, 23:37:51) [GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/local/lib/python311.zip', '/usr/local/lib/python3.11', '/usr/local/lib/python3.11/lib-dynload', '/app/lib/python3.11/site-packages']
>>>
(app) root@e8eadcbb1a86:/app# which python
/app/bin/python
(app) root@e8eadcbb1a86:/app# grep Products.CMFPlone constraints.txt
Products.CMFPlone==6.0.14
```

6.1.0rc1 plone-backend
```
root@ffa942febe9c:/app# python
Python 3.11.11 (main, Jan 14 2025, 10:54:20) [GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/local/lib/python311.zip', '/usr/local/lib/python3.11', '/usr/local/lib/python3.11/lib-dynload']
>>> exit
Use exit() or Ctrl-D (i.e. EOF) to exit
>>>
(app) root@ffa942febe9c:/app# python3
Python 3.11.11 (main, Jan 14 2025, 10:54:20) [GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>
(app) root@ffa942febe9c:/app# grep Products.CMFPlone constraints.txt
Products.CMFPlone==6.1.0rc1
```

fredvd committed:
use rc2 locally, but override server-builder dockerfile to fetch constraints.txt from rc1, add no-cache. Now build-images work locally.

fredvd commented:
@mauritsvanrees @davisagli @ericof  Found the issue, I think.  my local build was still fetching/preferring already built images from hub.docker.com.  

I've set the PLONE_VERSION to 6.1.0rc2  here, but have hardcoded 6.1.0rc2 in  Dockerfile.builder where it still fetches the 6.1.0rc1/constraints.txt from dist.plone.org. 

Should/can we remove the 6.1.0rc1 images from hub.docker.com? Then we can still release under rc1 later.

fredvd commented:
note: needs some feedback, fixes and cleanup before it can be merged

fredvd commented:
Oh, one more request from me, but can be dealt with in another issue/PR.  But now that I might have your attention here ;-) :

can we move/rename the /app folder in the backend images to /backend ?

this would make it much easier to build a single plone-demo image with both of them in. Dockerfile later

mauritsvanrees commented:
I don't completely follow what is happening here, I am not the greatest Docker guru.  But I see a lot of good things here.  +1 for moving to Python 3.12 and latest RelStorage.

fredvd commented:
And: we need to check and maybe mention an issue when migrating from Relstorage 3.5 to 4.0 or 4.x and ZODB issues . Maurits found the report back when I mentioned it to him this afternoon: 

https://github.com/zodb/relstorage/issues/505

davisagli commented on Dockerfile:
```diff
@@ -1,12 +1,18 @@
 LABEL maintainer="Plone Community <dev@plone.org>" \
       org.label-schema.name="plone-backend" \
-      org.label-schema.description="Plone backend image using Python $PYTHON_VERSION" \
+      org.label-schema.description="Plone $PLONE-VERSION backend image using Python $PYTHON_VERSION" \
```
```suggestion
      org.label-schema.description="Plone $PLONE_VERSION backend image using Python $PYTHON_VERSION" \
```

davisagli commented on Dockerfile:
```diff
@@ -1,12 +1,18 @@
 # syntax=docker/dockerfile:1
-ARG PYTHON_VERSION=3.11
+ARG PYTHON_VERSION=3.12
+ARG PLONE_VERSION=6.1.0rc2
```
This was previously passed in from the GHA based on version.txt, so we only needed to update the version in one place. Is there a reason that won't work any longer?

fredvd replied:
It is still passed in as --build-arg when we call buildx from the Makefile. 

I added the default values to silence a warning from buildx that a variable is used in the dockerfile that can be empty if it isn't passed in as a --build-arg :-P

davisagli replied:
@fredvd Well, we could just leave the warning there since we were doing it intentionally. I'm worried that we'll miss updating all copies of the version when we make updates in the future.

davisagli commented on Dockerfile.builder:
```diff
@@ -28,11 +28,11 @@ RUN <<EOT
     apt-get update
     apt-get -y upgrade
-    buildDeps="build-essential busybox ca-certificates curl git gosu libbz2-dev libffi-dev libjpeg62-turbo-dev libmagic1 libldap2-dev libopenjp2-7-dev libpcre3-dev libpq-dev libsasl2-dev libssl-dev libtiff5-dev libxml2-dev libxslt1-dev wget zlib1g-dev"
+    buildDeps="build-essential busybox ca-certificates curl git gosu libbz2-dev libffi-dev libjpeg62-turbo-dev libmagic1 libsasl2-dev libldap2-dev libopenjp2-7-dev libpcre3-dev libpq-dev libsasl2-dev libssl-dev libtiff6 libtiff5-dev libxml2-dev libxslt1-dev wget zlib1g-dev"
```
libsasl2-dev is now here twice. And, we need both libtiff6 and libtiff5-dev?

fredvd replied:
The duplication is an error, I'll fix it.  And yes, apparently we need both in bookworm. libtiff5 doesn't exist anymore in the distro, libtiff5-dev still is and  works with libtiff6 as a library, it is mentioned it is a transitional package.

davisagli replied:
The duplication of libsasl2-dev is still here

fredvd replied:
Just removed

davisagli commented on Dockerfile.classicui:
```diff
@@ -1,13 +1,18 @@
-FROM plone/server-builder:${PLONE_VERSION} as builder
 
-# Remove Volto support from builder image
+# Remove Volto support from builder image, this is Classic-UI
```
```suggestion
# Remove Volto support from builder image, this is Classic UI
```

davisagli commented on version.txt:
```diff
@@ -1 +1 @@
-6.1.0rc1
+6.1.0rc2
```
There is no https://dist.plone.org/release/6.1.0rc2 and we should really keep the image version matching there. I didn't follow why you used rc2 here?

fredvd replied:
I changed the version in version.txt and at some other places temporaily to rc2 as a test to check if I was dealing with caching issues and/or pulling in an already existing rc1 server-builder or server-prod-config from hub.docker.com.  

My local builds of plone-backend were broken with a mixed Python3.11/Python3.12 environment that didn't make sense.   I'll update the PR, remove the older ones from hub.docker.com and see if we can release the images now properly as rc1 with GHA

davisagli commented:
@fredvd 

> Should/can we remove the 6.1.0rc1 images from hub.docker.com? Then we can still release under rc1 later.

I don't see 6.1.0rc1 images there, so I guess you already did that? Yes, I'd say it's important to release under rc1 to match what is on dist.plone.org.

> can we move/rename the /app folder in the backend images to /backend ? this would make it much easier to build a single plone-demo image with both of them in. Dockerfile later

That sounds like a breaking change that would affect most downstream images that build on top of this, so I'm pretty strongly -1 on that. But maybe it could be an ARG? I would not do it pre-emptively until you actually try to create a unified image and see whether it makes sense to reuse these Dockerfiles or not.

fredvd committed:
Update Dockerfile

Co-authored-by: David Glick <david@glicksoftware.com>

fredvd committed:
Update Dockerfile.classicui

Co-authored-by: David Glick <david@glicksoftware.com>

fredvd committed:
Clean up rc2 tests: set back to rc1, revert hardcoded rc1 that fetches constraints.
Remove version nmber from PLONE_VERSION ARG in Docker files so that the build will fall. This will re-introduce output warnings from docker buildx that possible uninitialised variables are used.

fredvd committed:
Add warning in README on updates OS and library major versions in 6.1.x

davisagli commented on README.md:
```diff
@@ -31,6 +31,10 @@ These images are **not** Buildout based!
 
+**Possible breaking changes in upcoming Plone 6.1 dependencies**
+
+Please note that the backend images in the 6.1.x series have several updated dependencies, which you should verify and test before using them in any existing project. Even if you extend from these images. The underlying OS has been switched from Debian 11 buster to 12 bookworm. Relstorage has a new major release going from 3.x to 4.1. And there are are minor updates for libldap, libtiff and psyco ...
```
Mention the update to Python 3.12 also!

fredvd replied:
good catch.

fredvd committed:
Remove double libsaasl2-dev definition

fredvd commented:
> @fredvd
> 
> > Should/can we remove the 6.1.0rc1 images from hub.docker.com? Then we can still release under rc1 later.
> 
> I don't see 6.1.0rc1 images there, so I guess you already did that? Yes, I'd say it's important to release under rc1 to match what is on dist.plone.org.
> 
> > can we move/rename the /app folder in the backend images to /backend ? this would make it much easier to build a single plone-demo image with both of them in. Dockerfile later
> 
> That sounds like a breaking change that would affect most downstream images that build on top of this, so I'm pretty strongly -1 on that. But maybe it could be an ARG? I would not do it pre-emptively until you actually try to create a unified image and see whether it makes sense to reuse these Dockerfiles or not.

Ah, a variable to parameterise the docker-entrypoint.sh is a good suggestion, I can do so something with that, thanks!

fredvd committed:
Mention Python 3.12

fredvd commented:
I have removed the rc2 testing setup from the branch and removed the default ARG variables.  I have also removed the plone/server-prod-config 6.1.0rc1 tag from hub.docker.com as that was likely the cause why local test builds of the rc1 images had a mixed python3.11/3.12 environment. 

 If this PR is merged, rc1 should build and be deployed to hub.docker.com

mauritsvanrees commented:
I think the [old v6.1.0rc1 tag](https://github.com/plone/plone-backend/releases/tag/v6.1.0rc1) in this repo can be removed, and a new tag made, right?

fredvd commented:
> I think the [old v6.1.0rc1 tag](https://github.com/plone/plone-backend/releases/tag/v6.1.0rc1) in this repo can be removed, and a new tag made, right?

Yes, that archive/release you can download is a bit superfluous, because what counts are the released images. But the contents of the tar.gz source files should match the generated container image tags, which should match the Plone (backend) release.  

I'll remove the tag and will try to build/release the 6.1.0rc1 images again.

mauritsvanrees commented:
I don't really care about the archive/release on GitHub.  I only meant that a new tag was needed, because that is the only way I know that an image will be generated.
But I see docker images can be made by manually triggering the [workflow](https://github.com/plone/plone-backend/actions/workflows/release.yml).
